File: backend/src/main/java/com/mini/credit/service/EcartCaisseService.java
package com.mini.credit.service;

import com.mini.credit.dto.caisse.EcartCaisseDTO;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * PHASE 7: Service pour écarts caisse.
 */
public interface EcartCaisseService {

    /**
     * Crée un écart caisse détecté
     */
    EcartCaisseDTO detecterEcart(
            Long sessionCaisseId, Long recetteId, LocalDate dateJour, String typeEcart,
            BigDecimal montantEcart, String description, Boolean seuilDepassé);

    /**
    * PATCH 8 - Récupère tous les écarts (CONTROLEUR, CHEF_BUREAU, ADMIN)
     */
    List<EcartCaisseDTO> getAll();

    /**
     * Récupère un écart par ID
     */
    EcartCaisseDTO getById(Long id);

    /**
     * PATCH 8 — Récupère les écarts d'une session caisse
     */
    List<EcartCaisseDTO> getBySessionCaisseId(Long sessionId);

    /**
     * Récupère les écarts détectés pour un jour
     */
    List<EcartCaisseDTO> getByDateJour(LocalDate dateJour);

    /**
     * Récupère les écarts en investigation (CONTROLEUR)
     */
    List<EcartCaisseDTO> getEnInvestigation();

    /**
     * Récupère les écarts nécessitant validation R.C.I.
     */
    List<EcartCaisseDTO> getRequiringRCIValidation();

    /**
     * PATCH 8 — Justification initiale d'un écart.
     * Accessible par CAISSIER et CONTROLEUR.
     * Ajoute une note de justification sans changer le statut.
     * NOTE: La règle de longueur minimale (10 car.) est une validation technique provisoire.
     * Elle n'est pas définie dans les documents 3N fournis.
     */
    EcartCaisseDTO justifier(Long ecartId, String justification);

    /**
     * PHASE 7: Enquête d'un écart (CONTROLEUR)
     * Passe en EN_INVESTIGATION
     */
    EcartCaisseDTO enqueterEcart(Long ecartId, String notesInvestigation);

    /**
     * PHASE 7: Résout un écart (CONTROLEUR)
     * Passe en RESOLU avec raison
     */
    EcartCaisseDTO resoudreEcart(Long ecartId, String raisonResolution);

    /**
    * PHASE 7: Valide un écart (Chef de Bureau)
     * Passe en ACCEPTE (variance normalisée)
     */
    EcartCaisseDTO accepterEcart(Long ecartId);

    /**
     * Rejette un écart (présumé erreur)
     */
    EcartCaisseDTO rejeterEcart(Long ecartId);

    /**
     * Total écarts non résolus pour un jour
     */
    BigDecimal getTotalEcartsNonResolusByDateJour(LocalDate dateJour);
}

File: frontend/src/app/features/caisse/services/ecart-caisse.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../../../core/services/api.config';
import { EcartCaisseResponse } from '../models/ecart-caisse-response';
import { EcartCaisseJustifierRequest } from '../models/ecart-caisse-justifier-request';
import { EcartCaisseResoudreRequest } from '../models/ecart-caisse-resoudre-request';

/**
 * PATCH 9 — Service Angular pour la gestion des écarts de caisse.
 *
 * RBAC appliqué côté backend (@PreAuthorize). Ce service ne filtre pas les actions
 * selon le rôle — c'est la responsabilité des composants et du guard de route.
 *
 * Mapping rôles 3N → métier :
 *   Chef de Bureau  = supervision d'antenne
 *   Contrôleur      = contrôle caisse
 *   Caissier        = opérations journée caisse
 *
 * Seuils : aucun seuil fixe côté frontend.
 * Le flag seuilDepassé est fourni par le backend via EcartThresholdConfigService.
 */
@Injectable({
  providedIn: 'root'
})
export class EcartCaisseService {
  private readonly apiUrl = `${API_BASE_URL}/ecarts-caisse`;

  constructor(private http: HttpClient) {}

  // ===== LECTURE =====

  /** ADMIN, CONTROLEUR, CHEF_BUREAU — liste tous les écarts. */
  getAll(): Observable<EcartCaisseResponse[]> {
    return this.http.get<EcartCaisseResponse[]>(this.apiUrl);
  }

  /** ADMIN, CONTROLEUR, CHEF_BUREAU, CAISSIER — détail d'un écart. */
  getById(id: number): Observable<EcartCaisseResponse> {
    return this.http.get<EcartCaisseResponse>(`${this.apiUrl}/${id}`);
  }

  /** ADMIN, CONTROLEUR, CHEF_BUREAU, CAISSIER — écarts d'une session. */
  getBySession(sessionId: number): Observable<EcartCaisseResponse[]> {
    return this.http.get<EcartCaisseResponse[]>(`${this.apiUrl}/session/${sessionId}`);
  }

  /** ADMIN, CONTROLEUR, CHEF_BUREAU — écarts par date (format ISO 8601 : yyyy-MM-dd). */
  getByJour(dateJour: string): Observable<EcartCaisseResponse[]> {
    return this.http.get<EcartCaisseResponse[]>(`${this.apiUrl}/jour/${dateJour}`);
  }

  /** ADMIN, CONTROLEUR — écarts DETECTE ou EN_INVESTIGATION. */
  getEnqueteEnAttente(): Observable<EcartCaisseResponse[]> {
    return this.http.get<EcartCaisseResponse[]>(`${this.apiUrl}/enquete/en-attente`);
  }

  /** ADMIN, CHEF_BUREAU — écarts nécessitant validation hiérarchique (seuilDepassé=true). */
  getValidationHierarchique(): Observable<EcartCaisseResponse[]> {
    return this.http.get<EcartCaisseResponse[]>(`${this.apiUrl}/validation/hierarchique`);
  }

  /** ADMIN, CONTROLEUR, CHEF_BUREAU — total non résolu pour un jour (yyyy-MM-dd). */
  getTotalNonResoluByJour(dateJour: string): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/total-non-resolu/jour/${dateJour}`);
  }

  // ===== WORKFLOW =====

  /**
   * ADMIN, CONTROLEUR, CAISSIER — justification initiale (n'change pas le statut).
   *
   * NOTE TECHNIQUE : La validation 10 caractères minimum est provisoire.
   * La règle métier 3N impose uniquement qu'une justification soit documentée.
   */
  justifier(id: number, request: EcartCaisseJustifierRequest): Observable<EcartCaisseResponse> {
    return this.http.post<EcartCaisseResponse>(`${this.apiUrl}/${id}/justifier`, request);
  }

  /**
   * ADMIN, CONTROLEUR — ouvre une enquête formelle (statut → EN_INVESTIGATION).
   * Utilise le même payload que justifier (champ justification).
   */
  ouvrirEnquete(id: number, request: EcartCaisseJustifierRequest): Observable<EcartCaisseResponse> {
    return this.http.post<EcartCaisseResponse>(`${this.apiUrl}/${id}/ouvrir-enquete`, request);
  }

  /** ADMIN, CONTROLEUR — résolution avec raison documentée (statut → RESOLU). */
  resoudre(id: number, request: EcartCaisseResoudreRequest): Observable<EcartCaisseResponse> {
    return this.http.post<EcartCaisseResponse>(`${this.apiUrl}/${id}/resoudre`, request);
  }

  /** ADMIN, CHEF_BUREAU — accepte la variance (statut → ACCEPTE). */
  accepter(id: number): Observable<EcartCaisseResponse> {
    return this.http.post<EcartCaisseResponse>(`${this.apiUrl}/${id}/accepter`, {});
  }

  /** ADMIN uniquement — rejette l'écart (erreur système présumée, statut → REJETE). */
  rejeter(id: number): Observable<EcartCaisseResponse> {
    return this.http.post<EcartCaisseResponse>(`${this.apiUrl}/${id}/rejeter`, {});
  }
}
